iroh-ctl/src/run.rs

```diff
@@ -0,0 +1,142 @@
+use std::collections::HashMap;
+use std::env;
+use std::path::PathBuf;
+
+use crate::config::{Config, CONFIG_FILE_NAME, ENV_PREFIX};
+#[cfg(feature = "testing")]
+use crate::fixture::get_fixture_api;
+use crate::{
+    gateway::{run_command as run_gateway_command, Gateway},
+    p2p::{run_command as run_p2p_command, P2p},
+    store::{run_command as run_store_command, Store},
+};
+use anyhow::Result;
+use clap::{Parser, Subcommand};
+use iroh::{Api, Iroh};
+use iroh_metrics::config::Config as MetricsConfig;
+use iroh_resolver::resolver;
+use iroh_rpc_client::Client;
+use iroh_util::{iroh_config_path, make_config};
+
+#[derive(Parser, Debug, Clone)]
+#[clap(version, about, long_about = None, propagate_version = true)]
+pub struct Cli {
+    #[clap(long)]
+    cfg: Option<PathBuf>,
+    #[clap(long = "no-metrics")]
+    no_metrics: bool,
+    #[clap(subcommand)]
+    command: Commands,
+}
+
+impl Cli {
+    fn make_overrides_map(&self) -> HashMap<String, String> {
+        let mut map = HashMap::new();
+        map.insert("metrics.debug".to_string(), self.no_metrics.to_string());
+        map
+    }
+}
+
+#[derive(Subcommand, Debug, Clone)]
+enum Commands {
+    // status checks the health of the different processes
+    #[clap(about = "Check the health of the different iroh processes.")]
+    Status {
+        #[clap(short, long)]
+        /// when true, updates the status table whenever a change in a process's status occurs
+        watch: bool,
+    },
+    Version,
+    P2p(P2p),
+    Store(Store),
+    Gateway(Gateway),
+    #[clap(about = "break up a file into block and provide those blocks on the ipfs network")]
+    Add {
+        path: PathBuf,
+        #[clap(long, short)]
+        recursive: bool,
+        #[clap(long, short)]
+        no_wrap: bool,
+    },
+    #[clap(
+        about = "get content based on a Content Identifier from the ipfs network, and save it "
+    )]
+    Get {
+        path: resolver::Path,
+        #[clap(long, short)]
+        output: Option<PathBuf>,
+    },
+}
+
+impl Cli {
+    #[cfg(not(feature = "testing"))]
+    pub async fn run(&self) -> Result<()> {
+        self.run_impl().await
+    }
+
+    #[cfg(feature = "testing")]
+    pub async fn run(&self) -> Result<()> {
+        let api = get_fixture_api();
+        self.cli_command(&api).await
+    }
+
+    // extracted this into a public function so that we don't get a lot of
+    // rust analyzer unused code errors, which we do if we inline this code inside
+    // of run
```

Comment:
That's not how it is supposed to work. Why is it unused? if it is just used in testing, it should be disabled using the feature detection in the regular case

Reply:
also why is this not just inline with the above method?

Reply:
I started out having it just inlined.

Rust analyzer declares the version of `run` with the annotation `#[cfg(not(feature = "testing))"`
as unused. This is because the `testing` is enabled in dev mode due to the line in `Cargo.toml` I need to have in order to have `iroh-ctl` compile in testing mode. 

When I use `cargo run --bin iroh-ctl p2p peers` wants to talk to iroh over RPC, so this logic does get picked up.

The version of `run` with `#[cfg(feature = "testing))]"`  is not declared unused by Rust analyzer as we're in development mode in the editor and therefore `testing` is enabled.

I extracted the function `run_impl` as I wanted to have Rust analyzer help me instead of marking it all as unused. Now only the small invocation of `run_impl` is marked as unused. 

Without  `pub` for `run_impl`, the function itself becomes marked unused. This gives me the following build warnings:

```
  --> iroh-ctl/src/run.rs:33:8
   |
33 |     fn make_overrides_map(&self) -> HashMap<String, String> {
   |        ^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(dead_code)]` on by default

warning: associated function `run_impl` is never used
  --> iroh-ctl/src/run.rs:86:14
   |
86 |     async fn run_impl(&self) -> Result<()> {
```

I've tried to add some comments to explain this better now.

Reply:
I considered alternatives. This approach you gave me:

https://github.com/bazelbuild/rules_rust/discussions/913

I don't know how to control `rustc_flags` in Cargo, or how to have a special bin that's ideally called `iroh-ctl` (or `iroh` in the future) that's compiled in `test` mode. See also for the open issue:

https://github.com/rust-lang/cargo/issues/1982

I also looked at:

https://docs.rs/test-binary/latest/test_binary/

This requires a lot of project setup and assumes I build the test binary in the test. I also don't know how to get `trycmd` to use these binaries as they're in sub-projects.